### PR TITLE
Add Vadere trajectory loader

### PR DESCRIPTION
**Pull request by gitsimonr**
This feature allows users to load trajectory and scenario files that originate from Vadere. The jupyter notebook `user_guide` provides an example. Data for the example and test data are added. Currently, this feature supports data created with Vadere 3.0.

### CODE REVIEW

**Comment by schroedtert** [comment]
Thanks for your contribution, we are really looking forward to support Vadere trajectory files! 
I will try to take a look at your work next week, unfortunately I won't have the time before.

If you have any question or need some assistance, don't hesitate to ask. Especially when our contributing workflow is not clear (you are the first external contributor).

File: pedpy/io/trajectory_loader.py [hunk 1]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
...
+            dtype={
+                VADERE_KEY_ID: "int64",
+                VADERE_KEY_TIME: "float64",
+                VADERE_KEY_X: "float64",
+                VADERE_KEY_Y: "float64",
+            },
+            encoding="utf-8-sig",
+        )
+
+        data.rename(columns=rename_mapping, inplace=True)
```
Comment (schroedtert) [inline pedpy/io/trajectory_loader.py:800]: From the ruff linter: PD002 `inplace=True` should be avoided; it has inconsistent behavior

File: pedpy/io/trajectory_loader.py [hunk 2]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
...
+               * 'shape' POLYGON requires key value pair for 'points': [{'x': ..., 'y': ...},
+               {'x': ..., 'y': ...}, ...]
+
+        decimals: Integer defining the decimals of the returned coordinates
+
+    Returns:
+        list
+
+    """
+    _supported_types = ["RECTANGLE", "POLYGON"]
```
Comment (schroedtert) [inline pedpy/io/trajectory_loader.py:1014]: Does Vadere currently (or in the near future) support more shape types?

File: tests/unit_tests/io/test_trajectory_loader.py [hunk 3]
```diff
@@ -361,6 +364,99 @@ def write_viswalk_csv_file(
...
+    data = data.rename(
+        columns={
+            ID_COL: "pedestrianId",
+            FRAME_COL: "simTime",
+            X_COL: "startX-PID1",  # "-PID1" stands for processor id 1 and is used in Vadere
+            # outputs as extension of the generic column name startX
+            Y_COL: "startY-PID1",  # "-PID1" see comment above
+        }
+    )
+    data["simTime"] = data["simTime"] / frame_rate
```
Comment (schroedtert) [inline tests/unit_tests/io/test_trajectory_loader.py:382]: Wouldn't this be nan for the default `frame_rate` (=0). If so, is the default=0 really needed?

**Review by schroedtert — CHANGES_REQUESTED** [review]
This looks very good, I just have some points which I want to clarify, before merging this. If you could reply to these remarks, that would be great.

On a different note, you are a bit unfortunate, as we just changed our linting and formatting to ruff. Running the checks locally, yields around 40 places where (mostly minor) remarks are given. If you need any assistance in getting your state to the current main or dealing with the linter warnings just let me know. As I am not that much into the Vadere trajectory format, I'm trusting you, that you kept all the edge cases in mind and handled. And it's great that you added test cases for your changes!

Thanks again for contributing and this great PR. I am hopeful, that we will be able to merge this soon!

**Comment by chraibi** [comment]
@gitsimonr Any updates on this PR? We are preparing the next release and we would love lo have this PR accepted.

**Comment by gitsimonr** [comment]
> @gitsimonr Any updates on this PR? We are preparing the next release and we would love lo have this PR accepted.

Unfortunately, I was not yet able to continue my work. I'll try to address the requested changes as soon as possible.

**Comment by chraibi** [comment]
Welcome @﻿Hayato-Hess to this PR! :-)

File: pedpy/io/trajectory_loader.py [hunk 4]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
         return data
     except pd.errors.ParserError as exc:
         raise LoadTrajectoryError(common_error_message) from exc
+
+
+def load_trajectory_from_vadere(
+    *,
+    trajectory_file: pathlib.Path,
+    frame_rate: float = 24.0,
```
[the remark this answers is not in the document]
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:663]: From what I understand, no. Vadere simulates in 0.4-second intervals. Within each interval, depending on the model, a pedestrian may take multiple steps. The `frame_rate` parameter appears to control the precision of the interpolated steps in the output. A frame rate of 24 frames per second is roughly ten times the simulation frequency of Vadere. I assume the rate was chosen somewhat arbitrarily.

File: pedpy/io/trajectory_loader.py [hunk 5]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
...
+
+    max_deviation = round(max_speed / frame_rate, 2)
+    if max_deviation > deviation_threshold:
+        _log.warning(
+            f"The interpolated trajectory potentially deviates up to "
+            f"{str(max_deviation)} m from the original trajectory, at least "
+            f"for the fastest pedestrian with max. speed of {str(max_speed)} m/s. "
+            f"If smaller deviations are required, choose a higher frame rate. "
+            f"The current frame rate is {str(frame_rate)} fps."
+        )
```
[the remark this answers is not in the document]
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:927]: Personally, I think a warning could be a reasonable solution here. It would inform the user that, given the current `frame_rate`, the results might contain some inaccuracies. This way, the user can decide whether to tolerate the imprecision or adjust the settings accordingly.

To draw a perhaps imperfect analogy. When compressing music, the quality depends on the sampling rate: higher or lower, each has trade-offs. Yet, we still allow users the freedom to choose their preferred rate.

File: pedpy/io/trajectory_loader.py (hunk 2 above)
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:1014]: Currently, Rectangle and Polygons are supported for obstacles. As of now, there are no plans to add more shape types.

File: pedpy/io/trajectory_loader.py (hunk 1 above)
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:800]: It appears that there is an ongoing effort within the pandas library to remove inplace support for the rename method (see [PR #51466](https://github.com/pandas-dev/pandas/pull/51466/files/92c6a0a253858405b7a9552eb2bb2cf9fa1fcaf4#diff-f2cc21ad9c9caffa3506c4719e07e1db49a2a3368bd6b41a1abf90eb8e3e416cR11). Although the change is still open for discussion, the use of inplace seems to be quite controversial. 

I suggest replacing this line with:

    data = data.rename(columns=rename_mapping)

File: pedpy/io/trajectory_loader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (schroedtert) [reply pedpy/io/trajectory_loader.py:663]: If there is no actual meaning of a frame rate of 24 in Vadere, I would suggest removing the default value. Do you agree? Thus making it more transparent to the users that this needs to be passed to the function explicitly.

File: pedpy/io/trajectory_loader.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (schroedtert) [reply pedpy/io/trajectory_loader.py:927]: Agreed, logging a warning to the users seems to be a suitable solution here

File: pedpy/io/trajectory_loader.py [hunk 6]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
...
+        t_stop_ = (
+            math.floor(np.round(t_stop * frame_rate, precision)) / frame_rate
+        )
+
+        if t_start == t_stop:
+            _log.warning(
+                f"Trajectory of pedestrian {str(ped_id)} is too short in time "
+                f"to be captured by the chosen frame rate of {str(frame_rate)}. "
+                f"Therefore, this trajectory will be ignored."
```
[the remark this answers is not in the document]
Comment (schroedtert) [reply pedpy/io/trajectory_loader.py:851]: Actually, I like your idea of adding a `ignore_too_short_trajectories` parameter to the function. Its default should be 'False`, such that the function raises an error when there are too short trajectories. This should in general be an edge case, as the trajectories need to be really short to not be included in a reasonable frame rate. When the trajectory is ignored, I would still suggest printing a warning, making it visible to the user that some trajectory is ignored. 

I am not sure if feasible, but I think it would be great that in case of too short trajectory, a list with all trajectories that are too short is displayed together with a frame rate where they would be included. E.g.,
> Trajectory of pedestrians 1 (0.1s), 5 (0.2s), 8 (0.1s) are too short in time to be captured by the chosen frame rate of 25. To include these trajectories in the data please chose a frame rate of at least 50 or to ignore these trajectories set `ignore_too_short_trajectories=True`.

Just to sum up my preferred approach, if you have any points you disagree with, let me know. 

- add parameter `ignore_too_short_trajectories=False` to `load_trajectory_from_vadere`
- if `ignore_too_short_trajectories=False`:
  - raise errors when trajectories are to short. The error message should show a proposed frame rate, that would include the trajectory
- if `ignore_too_short_trajectories=True`:
  - ignore too short trajectories, but display a warning like: Trajectory X is ignored for frame rate Y

**Comment by schroedtert** [comment]
Hi @Hayato-Hess, 
thanks taking over this PR and for your feedback on my points. I really appreciate it and hope we will get this merged soon! 

As this PR has been open for quite a while, there might be some/more merge conflicts. And as I pointed out before, we switched to ruff as linter shortly before this PR got opened. Meanwhile, we set the linting/formatting up as pre-commit hooks. as described in our [Developer Guide](https://pedpy.readthedocs.io/stable/developer_guide.html#formatting-linting). If you need any assistance or anything is unclear, just let us know and we'll try to help you.

File: pedpy/io/trajectory_loader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:663]: I agree not to pass a default value. 

@chraibi I can not change the PR. See if you can force-push the changes. Otherwise I could also create a new PR.

**Comment by chraibi** [comment]
> @chraibi I'm unable to make changes to this PR. I could create a new branch with the changes we discussed above. We could then try to force-push the changes into this PR. If this doesn't work I could also open a new PR.

Is it maybe better to clone this branch and base your work on it?

File: pedpy/io/trajectory_loader.py [hunk 7]
```diff
@@ -651,3 +655,389 @@ def _load_trajectory_data_from_viswalk(
...
+def load_trajectory_from_vadere(
+    *,
+    trajectory_file: pathlib.Path,
+    frame_rate: float = 24.0,
+) -> TrajectoryData:
+    """Loads trajectory data from Vadere-traj file as :class:`~trajectory_data.TrajectoryData`.
+
+    This function reads a traj file containing trajectory data from Vadere simulations and
+    converts it into a :class:`~trajectory_data.TrajectoryData` object which can be used for
+    further analysis and processing in the *PedPy* framework.
```
[the remark this answers is not in the document]
Comment (Hayato-Hess) [reply pedpy/io/trajectory_loader.py:669]: I noticed that an error message for missing fields is already in place:
https://github.com/PedestrianDynamics/PedPy/pull/382/files/40800a90f21cbe11fa1ce6ced03d4c3905a9d59e#diff-f8408bffe4719804c18600918d9c725bb38ce03f631aee71166e14e7ba493995R737

I'll update the message to also suggest that the issue might be due to using an outdated Vadere version used for the export.

**Comment by Hayato-Hess** [comment]
@chraibi I created a draft PR https://github.com/Hayato-Hess/PedPy/pull/2 with the discussed changes. I hope you can grab the changes from there and force push them.

@schroedtert It would be great if you could check if the changes in https://github.com/Hayato-Hess/PedPy/commit/8b465a6d75a0e7762d3d4173f18bc756a1d3ee3b are what you envisioned

**Comment by schroedtert** [comment]
Superseded by #460
